src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs

```diff
@@ -231,7 +231,13 @@ async Task RenewMessageLockTask(Message message, CancellationToken renewLockCanc
                 {
                     var amount = MessagingUtilities.CalculateRenewAfterDuration(message.SystemProperties.LockedUntilUtc);
                     MessagingEventSource.Log.MessageReceiverPumpRenewMessageStart(this.messageReceiver.ClientId, message, amount);
-                    await Task.Delay(amount, renewLockCancellationToken).ConfigureAwait(false);
+                    var delayTask = await Task.Delay(amount, renewLockCancellationToken)
+                        .ContinueWith(t => t)
```

Comment:
By doing ```ContinueWith()``` we're not awaiting the canceled task (which would lead to the original TaskCanceledException), but we await a new task (which is _not_ canceled). The new task is completed immediately after ```Task.Delay()```is canceled (or completed if the delay time has elapsed) and return the delay task. Then we can check if the returned task (the delay task) is canceled or completed.

Reply:
Thank you @havarnov .

I was a bit confused by the code as one has to remember that the first part (`Task.Delay()`) is not what's awaited. Rather the continuation, which returns `Task<Task>` or `Task<Task.Delay()>`. Seeing `await` keyward in front is very confusing.

I would suggest to replace that code with a more explicit / intention revealing code that is equivalent to what you're trying to accomplish:

``` c#
var task = Task.Run(async () => 
   await Task.Delay(TimeSpan.FromSeconds(10), cts.Token), CancellationToken.None);
```

Reply:
I agree that the code might be a bit confusing. But I'm not sure if your proposal makes it less so. Because then you'd have to ```await``` the new task on the next line, like this:

```cs
var task = Task.Run(async () => 
   await Task.Delay(TimeSpan.FromSeconds(10), cts.Token), CancellationToken.None);
var delayTask = await task;
if (delayTask.IsCanceled) { ... }
```

If you still think this is better I'll change it, but maybe just a extensive comment would suffice?

Reply:
Both the options look good.. 
I personally feel @SeanFeldman 's code looks easier to read, but nothing a comment cannot solve.

Reply:
I also like @SeanFeldman's version more 😀

Reply:
Ok, then I'll change it to @SeanFeldman's version and add this comment:

> We're awaiting the task created by Task.Run() to avoid awaiting the Delay task which may be canceled by the renewLockCancellationToken. This way we prevent a TaskCanceledException.

ok?

Reply:
```cs
var task = Task.Run(
    async () =>
    {
        var innerDelayTask = Task.Delay(TimeSpan.FromSeconds(10), renewLockCancellationToken);
        await innerDelayTask;
        return innerDelayTask;
    },
    CancellationToken.None);
var delayTask = await task;
if (delayTask.IsCanceled) { ... }
```

In my mind this is much more confusing than my version. Maybe I'm missing something?

Reply:
@havarnov see my explanation [here](https://github.com/Azure/azure-service-bus-dotnet/pull/548#discussion_r208731547). I asked a co-worker to review both options and he leaned toward version where he can see that task is kicked off and result is queries later. The challenge with Continuation is that you have to remember that what's awaited is not the first part (`Task.Delay`), but the continuation. Which I suspect is not very natural when you read left to right.

Reply:
@SeanFeldman @nemakam - the solution you propose still throws the exception, so then the code is exactly the same as it was before except that we have create one extra task.

I really can't come up with any other code that behaves the way we want (not throwing the TaskCanceledException) except the use of ```ContinueWith```.

```cs
var task = Task.Run(async () => 
   await Task.Delay(TimeSpan.FromSeconds(10), cts.Token), CancellationToken.None);
var delayTask = await task;
if (delayTask.IsCanceled) { ... }
```
In the previous snippet we still awaits the ```Task.Delay()``` task so the exception is still thrown.

But as I said, maybe I'm missing something?

Reply:
@SeanFeldman You misunderstood me. It's identical with the code _before_ this PR (except for one extra task).

What I'm saying is that the version with ```Task.Run()``` will still throw ```TaskCanceledException``` as before. And If you pass in ```cts.Token``` two canceled tasks will be awaiting and both throw.

To summarize: The only version I can think of that doesn't throw the ```TaskCanceledException``` is the current one.

Reply:
@havarnov I took some time to understand what's going on, asking people more knowledgeable than myself about this and found... magic 🙂 

### Analysis

#### `try/catch`

This is probably the most intuitive approach. You have a task that can be cancelled within a `try` block and `OpreationCancelledException` will be raised. That's according to the language spec and very simple to see/understand.

#### Task.Run(child task)

This one will still throw `OpreationCancelledException` when a child task is cancelled. That's by definition is the right behaviour: exceptions will bubble-up.

#### ContinueWith()

This one is a bit tricky. While there's a "child" task representing `Task.Delay` that get cancellend, it looks like continuation API is handling exceptions internally and **doesn't** bubble those up. While I don't have an exact answer why this is happening, I suspect this is due to the fact that `ContinueWith (task => task)` can be executed conditionally, based on the outcome of the original (child) task. E.g. `ContinueWith(t => t, TaskContinuationOptions.OnlyOnFaulted)`. Meaning, it _has_ to handle the exception rather than bubbling it up. Therefore using `ContinueWith()` swallows `OperationCancelledException`.

### Conclusion

Personally, I'd rather see `try/catch` with `OperationCancelledException`, not a general `Exception` the way it's handled in the [original code](https://github.com/Azure/azure-service-bus-dotnet/pull/548/files#diff-222db52862b453d922cecebfd07b7597L254). It's simple and allows every .NET developer understand what it stands for:
```c#
try
{
  await Task.Delay(time);
}
catch (OperationCancelledException)
{
  // operation cancelled, nothing to do
}
```

If @nemakam votes against that, then continuation approach would work, but this is not as evident and clear as the option I'd chose.

Reply:
It's interesting that Microsoft recommends ```try/catch``` for control flow. In some cases I agree that ```try/catch``` will be a good solution for ```Task.Delay```, but in many cases I think there are better solutions (out of scope for this discussion). For this particularly case I _don't_ think that ```try/catch``` is a good solution because it's in the hot path of the message handling and that the delay task will be canceled in most cases. 

Anyways, I think you're wrong in your paragraph about ```ContinueWith```. If you look at the implementation of ```Task.Delay``` (see [Task.cs in dotnet/coreclr](https://github.com/dotnet/coreclr/blob/4e89ab7c4011e927e2ff552435df56a77880d70f/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5390)) there is no exception thrown when the task is delayed. ```ContinueWith``` will handle the task internally with the async state machine and no exception is used internally for control flow. The ```TaskCanceledException``` comes from ```await canceledTask;``` alone, because a canceled task is awaited. That means that the version with ```ContinueWith``` is fundamentally different and will _not_ throw any exception.

Reply:
@havarnov I did not say `Task.Delay` is the one that throwing. What I did say was
> it looks like continuation API is handling exceptions internally and doesn't bubble those up

`ContinueWith(t=>t)` cannot proceed w/o the inner task to be completed/cancelled/faulted. Therefore it's the continuation that is handling inner task exception. Which it does I think [here](https://referencesource.microsoft.com/#mscorlib/system/threading/Tasks/TaskContinuation.cs,360), swallowing the exceptions and not propagating any. So it's still on a hot path like you said, just swallowed for you rather than you do it explicitly with `try/catch`. And that's what my argument is all about. To keep code simple as possible without hidden gotchas. In case I'm completely wrong about exception handling in this case, will be glad to expand my knowledge and see explanation on how it works. If it ends up to be like I said, explicit vs implicit exception handling, my vote goes to the option that help understand the code. And, the team will overrule any decision as they are the ones to support it long term 🙂

Reply:
Mye bad for explaining myself badly/incorrectly. I still do think that the continuation of the task handles the control flow without any exceptions. As I said the continuation will handle the async state machine internally.

If you run ContinueWith in VS you won't see any exceptions in the diagnostics tool.

PS.: Sorry for arguing so much about this. I do find this interesting and would like to understand this fully. Maybe I should open a question on stackoverflow if we don't get to the bottom of this here?

Reply:
The answer on SO thread is interesting. It shows some performance improvement with continuation and an additional optimization that can be applied using `TaskContinuationOptions.ExecuteSynchronously`. But the true answer I suppose is in profiling. I.e. how much of an advantage this code will yeild when profiled. Are we optimizing a tenth of percent or something more substantial.

Reply:
@SeanFeldman  Thanks for posting that on SO. The response was insightful.
@havarnov  I do think its a very tiny optimization, but I am personally in favor of this optimization unless it makes the code too difficult to read. With the help of the comment, I do not think the code gets difficult enough so as to close a PR. Lets go ahead with this change. Can you please add the `TaskContinuationOptions.ExecuteSynchronously` parameter as well and we can take this forward.